syntax = "proto3";

option java_multiple_files = true;
option java_package = "io.grpc.examples.helloworld";
option java_outer_classname = "HelloWorldProto";
option objc_class_prefix = "HLW";


package helloworld;

//python -m grpc_tools.protoc -I./grpc --python_out=. --grpc_python_out=. ./grpc/messages.proto

// The greeting service definition.
service Greeter {

  rpc GetAvailableModelTypes (HelloRequest) returns (AvailableModelTypes) {}

  rpc GetMLModelsList (HelloRequest) returns (MLModelsList) {}

  rpc CreateModel (CreateModelRequest_2) returns (ResponseCreatedModel){}

  rpc FitModel (stream SendFile) returns (ResponseFitModel){}

  rpc PredictData (stream SendFile) returns (ResponsePredictData){}

  rpc GetModelInfo (ModelName) returns (ResponseModelInfo){}

  rpc DeleteModel (ModelName) returns (MLModelsList){}
}

message HelloRequest {
  string name = 1;
}

message ModelName{
  string model_name = 1;
}

message AvailableModelTypes {
  repeated string available_model_types = 1;
}

message MLModelsList {
  repeated string ml_models_list = 1;
  int32 error_code = 2;
  string error_message = 3;
}
//
//message CreateModelRequest {
//  string model_type = 1;
//  float alpha = 2;
//  bool fit_intercept = 3;
//  bool normalize = 4;
//  int32 max_iter = 5;
//  int32 random_state = 6;
//  string splitter = 7;
//  int32 max_depth = 8;
//  int32 min_samples_split = 9;
//  int32 min_samples_leaf = 10;
//  float min_weight_fraction_leaf = 11;
//  string criterion = 12;
//}

message CreateModelRequest_2 {
  string model_type = 1;
  string model_params = 2;
}

message ResponseCreatedModel {
  string model_path = 1;
  string model_type = 2;
}

message FitMetaData {
  string extension = 1;
  string model_name = 2;
  string target_column = 3;
}

message PredictMetaData {
  string extension = 1;
  string model_name = 2;
}

message SendFile {
  oneof request {
    FitMetaData fit_metadata = 1;
    bytes chunk_data = 2;
    PredictMetaData predict_metadata = 3;
  }
}

message ResponseFitModel {
  bool is_trained = 1;
  int32 error_code = 2;
  string error_message = 3;
}

message ResponsePredictData {
  repeated float predict = 1;
  int32 error_code = 2;
  string error_message = 3;
}

message ResponseModelInfo {
  string model_params = 1;
  int32 error_code = 2;
  string error_message = 3;
}